src/workerd/api/headers.h

```diff
@@ -157,62 +154,63 @@ class Headers final: public jsg::Object {
 
   JSG_SERIALIZABLE(rpc::SerializationTag::HEADERS);
 
-  void visitForMemoryInfo(jsg::MemoryTracker& tracker) const {
-    for (const auto& entry : headers) {
-      tracker.trackField(entry.first, entry.second);
-    }
-  }
+  void visitForMemoryInfo(jsg::MemoryTracker& tracker) const;
+
+  // A header is identified by either a common header ID or an uncommon header name.
+  // The header key name is always identifed in lower-case form, while the original
+  // casing is preserved in the actual Header struct to support case-preserving display.
+  // TODO(perf): We can likely optimize this further by interning uncommon header names
+  // so that we avoid repeated allocations of the same uncommon header name. Unless
+  // it proves to be a performance problem, however, we can leave that for future work.
+  using HeaderKey = kj::OneOf<uint, kj::String>;
 
 private:
-  struct Header {
-    jsg::ByteString key;   // lower-cased name
-    jsg::ByteString name;
-
-    // We intentionally do not comma-concatenate header values of the same name, as we need to be
-    // able to re-serialize them separately. This is particularly important for the Set-Cookie
-    // header, which uses a date format that requires a comma. This would normally suggest using a
-    // std::multimap, but we also need to be able to display the values in comma-concatenated form
-    // via Headers.entries()[1] in order to be Fetch-conformant. Storing a vector of strings in a
-    // std::map makes this easier, and also makes it easy to honor the "first header name casing is
-    // used for all duplicate header names" rule[2] that the Fetch spec mandates.
-    //
-    // See: 1: https://fetch.spec.whatwg.org/#concept-header-list-sort-and-combine
-    //      2: https://fetch.spec.whatwg.org/#concept-header-list-append
-    kj::Vector<jsg::ByteString> values;
-
-    explicit Header(jsg::ByteString key, jsg::ByteString name,
-                    kj::Vector<jsg::ByteString> values)
-        : key(kj::mv(key)), name(kj::mv(name)), values(kj::mv(values)) {}
-    explicit Header(jsg::ByteString key, jsg::ByteString name, jsg::ByteString value)
-        : key(kj::mv(key)), name(kj::mv(name)), values(1) {
-      values.add(kj::mv(value));
+  struct Header final {
+    kj::Maybe<kj::String> name;
```

Comment:
I guess this is only filled in if it differs from the lower-cased key? Can you add a comment explaining that?